File: R/example15_4.R
# Example 15.4: GARCH(1,1) Model of DM/BP Exchange Rate
# Bollerslerv and Ghysels [1996], JBES, 307-327.
#
# daily exchange rate data from 1/3/1984 to 12/31/1991 (obs.=1974)
dmbp<-read.table("http://web.pdx.edu/~crkl/ceR/data/dmbp.txt",nrows=1974)
xrate<-as.ts(dmbp[,1])
friday<-as.ts(dmbp[,2])

lm.2<-lm(xrate~friday)
summary(lm.2)

lm.1<-lm(xrate~1)
summary(lm.1)
e<-lm.1$residuals

for(j in 1:5) print(Box.test(e,lag=j,type="Box-Pierce"))
for(j in 1:5) print(Box.test(e,lag=j,type="Ljung-Box"))
acf(e,type="correlation")
acf(e,type="partial")

archlmtest <- function (x, lags, demean = FALSE) 
{ 
  x <- as.vector(x) 
  if(demean) x <- scale(x, center = TRUE, scale = FALSE) 
  lags <- lags + 1 
  mat <- embed(x^2, lags) 
  arch.lm <- summary(lm(mat[, 1] ~ mat[, -1])) 
  STATISTIC <- arch.lm$r.squared * length(resid(arch.lm)) 
  names(STATISTIC) <- "Chi-squared" 
  PARAMETER <- lags - 1 
  names(PARAMETER) <- "df" 
  PVAL <- 1 - pchisq(STATISTIC, df = PARAMETER) 
  METHOD <- "ARCH LM-test" 
  result <- list(statistic = STATISTIC, parameter = PARAMETER, 
                 p.value = PVAL, method = METHOD, data.name = 
                   deparse(substitute(x))) 
  class(result) <- "htest" 
  return(result) 
} 
archlmtest(e,24)[1:3]

e2<-e^2
for(j in 1:5) print(Box.test(e2,lag=j,type="Box-Pierce"))
for(j in 1:5) print(Box.test(e2,lag=j,type="Ljung-Box"))
acf(e2,type="correlation")
acf(e2,type="partial")
shapiro.test(e2)

hist(e2,freq=F)
lines(density(e2))
qqnorm(e2)
qqline(e2)

library(rugarch)
mean1<-list(armaOrder=c(0,0),include.mean=TRUE)
var1<-list(model="sGARCH",garchOrder=c(1,1))
spec1<-ugarchspec(variance.model=var1,mean.model=mean1)
garch1<-ugarchfit(spec1,xrate)
garch1
plot(garch1)

rm(list=ls())

File: python/example15_4.py
# Example 15.4: GARCH(1,1) Model of DM/BP Exchange Rate
# Bollerslerv and Ghysels [1996], JBES, 307-327.
#
# daily exchange rate data from 1/3/1984 to 12/31/1991 (obs.=1974)
import numpy as np
import pandas as pd
import statsmodels.api as sm

dmbp = pd.read_csv("http://web.pdx.edu/~crkl/ceR/data/dmbp.txt",sep='\s+',
                   header=None,names=['xrate','friday'],nrows=1974)

lm2 = sm.OLS.from_formula('xrate~friday',data=dmbp).fit()
lm2.summary()
lm1 = sm.OLS.from_formula('xrate~1',data=dmbp).fit()
lm1.summary()

e = lm1.resid
e2 = e**2

import statsmodels.tsa.api as tsa
import statsmodels.graphics.tsaplots as tsaplt

# time series analysis: identification
tsaplt.plot_acf(e, zero=False, lags=30)
tsaplt.plot_pacf(e, zero=False, lags=30)
tsaplt.plot_acf(e2, zero=False, lags=30)
tsaplt.plot_pacf(e2, zero=False, lags=30)

# https://arch.readthedocs.io/en/latest/univariate/introduction.html
from arch import arch_model

# GARCH(1,1), a common model
# default model: p=1,q=1, normal distribution
garch1 = arch_model(dmbp.xrate).fit()
print(garch1.summary())
garch1.plot()
garch1.arch_lm_test(lags=10,standardized=True)

# Other modelspecifications are possible:
# GJR-GARCH(1,1) model, set o=1
garch2 = arch_model(dmbp.xrate,p=1,o=1,q=1).fit()
print(garch2.summary())
garch2.plot()
# TARCH/ZARCH model, set power=1
garch3 = arch_model(dmbp.xrate,p=1,o=1,q=1,power=1).fit()
print(garch3.summary())
garch3.plot()
# using Stdent's t distribution
garch4 = arch_model(dmbp.xrate,dist='StudentsT').fit()
print(garch4.summary())
garch4.plot()
